### PR TITLE
[Local Testing] SNOW-875801 Support Column.cast

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/mock/analyzer.py [hunk 1]
```diff
@@ -288,12 +289,12 @@ def analyze(
...
 
         if isinstance(expr, Literal):
-            sql = str(expr.value)
+            sql = to_sql(expr.value, expr.datatype)
             if parse_local_name:
                 sql = sql.upper()
             # single quote literal when called by function/method to describe the column name of a dataframe
             # this is to align with the behavior of snowpark python when running against snowflake.
-            return f"'{sql}'"
+            return f"{sql}"
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/analyzer.py:297]: I removed this to pass `test_cast_nan_column_name` but the change seems to not break any tests. Is this expected? @sfc-gh-aling

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/mock/analyzer.py:297]: For my education, what was the purpose of adding the single quotes before?

File: src/snowflake/snowpark/mock/analyzer.py [hunk 2]
```diff
@@ -528,7 +525,7 @@ def binary_operator_extractor(
         parse_local_name=False,
         escape_column_name=False,
     ) -> str:
-        operator = expr.sql_operator.lower() if not isinstance(expr, EqualTo) else "=="
+        operator = expr.sql_operator.lower() if not isinstance(expr, EqualTo) else "="
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/analyzer.py:528]: Do we need this line? `EqualTo` sql_operator is already `=`

```
class EqualTo(BinaryArithmeticExpression):
    sql_operator = "="
```

File: src/snowflake/snowpark/mock/functions.py [hunk 3]
```diff
@@ -242,12 +252,20 @@ def mock_to_date(column: ColumnEmulator, fmt: Union[ColumnEmulator, str] = None)
...
+        try:
+            if auto_detect and data.isnumeric():
+                res.append(
+                    datetime.datetime.utcfromtimestamp(
+                        process_numeric_time(data)
+                    ).date()
+                )
+            else:
+                res.append(datetime.datetime.strptime(data, date_format).date())
+        except BaseException:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/functions.py:264]: Is there a more specific exception?

File: src/snowflake/snowpark/mock/analyzer.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py:297]: this is to add single quote around a literal, e.g., `select colname where colname like 'word'`
I think `to_sql` (which calls `str_to_sql`) handles the literal case.

File: src/snowflake/snowpark/mock/functions.py [hunk 4]
```diff
@@ -396,6 +448,135 @@ def mock_to_timestamp(column: ColumnEmulator, fmt: Union[ColumnEmulator, str] =
...
+    elif isinstance(source_datatype, _NumericType):
+        if fmt:
+            raise NotImplementedError(
+                "[Local Testing] Use format strings with Numeric types in TO_CHAR is not supported yet."
+            )
+        func = partial(try_convert, lambda x: str(x), try_cast)
+    else:
+        func = partial(try_convert, lambda x: str(x), try_cast)
+    new_col = column.apply(func)
+    new_col.sf_type = ColumnType(StringType(), column.sf_type.nullable)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/functions.py:495]: I encounter one issue when implementing the pandas related feature.
can you check if the length of a StringType() needs to be set to the default `16777216` when to_char is called

File: src/snowflake/snowpark/mock/plan.py [hunk 5]
```diff
@@ -1031,8 +1037,47 @@ def calculate_expression(
         return result
     if isinstance(exp, Cast):
         column = calculate_expression(exp.child, input_data, analyzer, expr_to_alias)
-        column.sf_type = ColumnType(exp.to, exp.nullable)
-        return column
+        if isinstance(exp.to, DateType):
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/plan.py:1040]: not blocking the PR, there is another way to do is to maintain a map
```python
cast_type_convert_func = {
    DateType: "to_date"
    # more
convert_func_name = cast_type_convert_func[exp.to]
return _MOCK_FUNCTION_IMPLEMENTATION_MAP[convert_func_name](column, try_cast=exp.try_)

}
```

File: src/snowflake/snowpark/mock/functions.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py:264]: I'm also curious what will be the error raised here

File: src/snowflake/snowpark/mock/analyzer.py [hunk 6]
```diff
@@ -288,12 +289,12 @@ def analyze(
             return expr.sql
 
         if isinstance(expr, Literal):
-            sql = str(expr.value)
+            sql = to_sql(expr.value, expr.datatype)
             if parse_local_name:
                 sql = sql.upper()
             # single quote literal when called by function/method to describe the column name of a dataframe
             # this is to align with the behavior of snowpark python when running against snowflake.
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/analyzer.py:296]:
```suggestion
```

**Comment by sfc-gh-aling** [comment]
todo: for unsupported case, raise proper error with message it's not supported and asking user to implement their own by `@patch`.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**